### PR TITLE
Calibration improvements

### LINKED ISSUE #534: Consolidate weighting categories by unit
Household populations, benefit unit populations, populations and financial amounts.

### LINKED ISSUE #535: Add input variables to weighting procedure
`employment income`, etc.

### PR DESCRIPTION
**Pull request by nikhilwoodruff**
Implements remaining re-weighting procedure changes.

- [x] Category reorganisation and bug fixes
- [x] Input variables included in targets

### CODE REVIEW

**Comment by nikhilwoodruff** [comment]
Before this is merged, I'll update https://github.com/PolicyEngine/openfisca-uk-reweighting with the new training logs.

File: openfisca_uk/calibration/losses/categories/budgetary_impact/of_income_tax_by_income_band.py [hunk 1]
```diff
@@ -4,19 +4,22 @@
...
         year,
     ):
-        brackets = IncomeTaxRevenueByIncome.parameter_folder.brackets[
-            :-3
-        ]  # Last-3 brackets are impossible due to FRS non-capture
+        brackets = (
+            BudgetaryImpactOfIncomeTaxByIncomeBand.parameter_folder.brackets[
+                :-3
+            ]
+        )  # Last-3 brackets are impossible due to FRS non-capture
```
Comment (MaxGhenis) [inline openfisca_uk/calibration/losses/categories/budgetary_impact/of_income_tax_by_income_band.py:20]: Do we need to drop them? Should produce the same result either way right?

Keeping them would allow for more directly demonstrating how future revisions to improve it (synthesis) improve loss, without having to make the loss function incomparable to previous openfisca-uk versions.

File: openfisca_uk/calibration/losses/loss_calculator.py [hunk 2]
```diff
@@ -23,28 +17,17 @@ def __init__(self, sim: Microsimulation, validation_split: float = 0.1):
...
+        self.losses = [Households, Populations, Families, BudgetaryImpact]
         self.validation_split = validation_split
         self.sim = sim
         self.training_log = []
         self.metrics = sum(
             [list(loss.get_metric_names()) for loss in self.losses], []
         )
         non_population_metrics = sum(
-            [list(loss.get_metric_names()) for loss in self.losses[5:]], []
+            [list(loss.get_metric_names()) for loss in self.losses[2:]], []
```
Comment (MaxGhenis) [inline openfisca_uk/calibration/losses/loss_calculator.py]: What does the `2` correspond to? Could you make it a named constant for documentation?

File: openfisca_uk/calibration/losses/loss_calculator.py [hunk 3]
```diff
@@ -98,3 +81,24 @@ def compute_loss(
...
+            k (int, optional): The number of folds (results in validation splits of 1 / k). Defaults to 1.
+
+        Returns:
+            List[LossCalculator]: The list of loss calculators.
+        """
+        base_calculator = LossCalculator(sim, validation_split=1)
+        loss_calculators = [LossCalculator(sim) for _ in range(k)]
+        for calculator in loss_calculators:
+            calculator.validation_metrics = []
+        for metric_name in base_calculator.validation_metrics:
```
Comment (MaxGhenis) [inline openfisca_uk/calibration/losses/loss_calculator.py:109]: Could you add a comment asserting or describing how this assigns each metric to exactly one validation split?

File: openfisca_uk/calibration/losses/loss_category.py [hunk 4]
```diff
@@ -1,8 +1,16 @@
...
+from numpy.typing import ArrayLike
+from itertools import chain
+from functools import reduce
+from typing import Callable, Iterable, List, Tuple, Type
+from openfisca_uk import Microsimulation
+import tensorflow as tf
+import logging
+import os
+
+os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
```
Comment (MaxGhenis) [inline openfisca_uk/calibration/losses/loss_category.py]: What does this do?

File: openfisca_uk/calibration/losses/loss_category.py [hunk 5]
```diff
@@ -13,9 +21,18 @@ def weighted_squared_relative_deviation(
     return ((pred / actual) - 1) ** 2 * actual
 
 
+def weighted_squared_log_relative_deviation(
+    pred: tf.Tensor, actual: ArrayLike
+) -> tf.Tensor:
```
Comment (MaxGhenis) [inline openfisca_uk/calibration/losses/loss_category.py:30]: Docstring please

File: openfisca_uk/parameters/calibration/budgetary_impact/by_income_source.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+description: Income totals by source for UK households.
+metadata:
+  uprating:
+    parameter: uprating.earnings
+  propagate_metadata_to_children: true
+  reference:
+    - title: Survey of Personal Incomes (SPI) | National Statistics
+      href: https://www.gov.uk/government/statistics/earned-income-2010-to-2011
```
Comment (MaxGhenis) [inline openfisca_uk/parameters/calibration/budgetary_impact/by_income_source.yaml:9]:
```suggestion
      # Data is more recent than the URL suggests.
      href: https://www.gov.uk/government/statistics/earned-income-2010-to-2011
```

**Review by MaxGhenis — APPROVED** [review]
This is excellent code! Just left a few minor suggestions, mostly for comments. I don't know what the issue is with the notebook but can look more tomorrow.

File: openfisca_uk/calibration/losses/loss_category.py (hunk 4 above)
Comment (nikhilwoodruff) [reply openfisca_uk/calibration/losses/loss_category.py]: It's a failed attempt to fix https://stackoverflow.com/questions/35911252/disable-tensorflow-debugging-information, nothing functional so will remove.

File: openfisca_uk/calibration/losses/loss_calculator.py (hunk 2 above)
Comment (nikhilwoodruff) [reply openfisca_uk/calibration/losses/loss_calculator.py]: Yep, and changed to a more verbose indexing method by using the actual category names.

File: openfisca_uk/calibration/losses/categories/budgetary_impact/of_income_tax_by_income_band.py (hunk 1 above)
Comment (nikhilwoodruff) [reply openfisca_uk/calibration/losses/categories/budgetary_impact/of_income_tax_by_income_band.py:20]: True, changed.

**Comment by nikhilwoodruff** [comment]
Don't really understand why the documentation is failing to build, it's passing locally even on a blank environment.

**Comment by MaxGhenis** [comment]
Do you think it might be a jupyterbook issue? Wondering if pinning an earlier version could help.

**Comment by nikhilwoodruff** [comment]
> Do you think it might be a jupyterbook issue? Wondering if pinning an earlier version could help.

Maybe? Just added a step to the action to print out the log file, will see what that reveals.

**Comment by nikhilwoodruff** [comment]
Finally, found it (the issue was that `gif` needed the `plotly` extension in `setup.py`).
